### ISSUE TITLE
False negative on while_let_on_iterator

### ISSUE THREAD

**Issue by llogiq** [rust-lang-nursery/rust-clippy#1881]
The last change to the `while_let_on_iterator` lint made it stop linting `while let` loops within other loops. This is obviously a too broad measure, as seen by the following example:

```
loop {
    let i = get_iter();
    while let Some(x) = i.next() { .. }
}
```

This should be a `for` loop, because the iterator is defined *within* the outer loop. So the way to go here is to first look up the definition of the iterator, walk up the parent to its defining scope, then walk up from the loop. If we find the iterator defining scope, lint. If we meet a loop before finding it, bail.

Once we have this, we can also rip out the loop_count again.
